shell/app/modules/msp/monitor/trace-insight/pages/trace-querier/trace-graph/index.tsx

```diff
@@ -50,11 +55,49 @@ export function TraceGraph(props: IProps) {
   const [spanStartTime, setSpanStartTime] = React.useState(null! as number);
   const [timeRange, setTimeRange] = React.useState([null!, null!] as number[]);
   const [selectedSpanId, setSelectedSpanId] = React.useState(null! as string);
+  const [view, setView] = React.useState('waterfall');
   const spanData = getSpanEvents.useData();
   const spanDataSource = spanData?.spanEvents || [];
   const duration = max - min;
   const allKeys: string[] = [];
   const { serviceAnalysis } = (spanDetailData as MONITOR_TRACE.ISpanRelationChart) || {};
+  const [flameRef, { width: flameWidth }] = useMeasure();
+
+  const containerRef = React.useRef(null);
+  const [tooltipState, setTooltipState] = React.useState(
+    null as { content: MONITOR_TRACE.FlameChartData; mouseX: number; mouseY: number } | null,
+  );
+
+  function getMousePos(
+    relativeContainer: { getBoundingClientRect: () => DOMRect } | null,
+    mouseEvent: { clientX: number; clientY: number },
+  ) {
+    if (relativeContainer !== null) {
+      const rect = relativeContainer.getBoundingClientRect();
+      const mouseX = mouseEvent.clientX - rect.left;
+      const mouseY = mouseEvent.clientY - rect.top;
+      return { mouseX, mouseY };
+    } else {
+      return { mouseX: 0, mouseY: 0 };
+    }
+  }
+
+  const onMouseOver = (event: { clientX: number; clientY: number }, data: MONITOR_TRACE.FlameChartData) => {
+    const { name, value, children, serviceName, selfDuration, spanKind, component } = data;
+    setTooltipState({
+      content: { name, value, children, serviceName, selfDuration, spanKind, component },
+      ...getMousePos(containerRef.current, event),
+    });
```

Comment:
Any mouse move will trigger rerender, use ref instead of setState. Or use css pseudo-class instead of js calcuation.